src/hotspot/share/runtime/mutex.cpp

```diff
@@ -137,14 +137,14 @@ void Mutex::lock_without_safepoint_check() {
 
 
 // Returns true if thread succeeds in grabbing the lock, otherwise false.
-
+// Checking for a NULL owner avoids the call into platform code and hides the
+// potential difference in recursive locking behaviour on some platforms.
 bool Mutex::try_lock() {
   Thread * const self = Thread::current();
   // Some safepoint_check_always locks use try_lock, so cannot check
   // safepoint state, but can check blocking state.
   check_block_state(self);
-  if (_lock.try_lock()) {
-    assert_owner(NULL);
+  if (_owner == NULL && _lock.try_lock()) {
```

Comment:
It's not obvious that's it is correct to check == NULL. Couldn't there be a lingering value from another thread, causing a spurious failure to acquire the lock? Could this be changed to _owner != self? That way we only check for a value that only the current thread could have written.

Also, this is a racy read, should this be using Atomic::load?

Reply:
I agree with the switch to checking `_owner != self`. The ObjectMonitor
code does checks in the same way because a thread should only check
for a value that it could have written.